src/draw/sw/lv_draw_sw_gradient.c

```diff
@@ -148,134 +156,154 @@ static lv_res_t kill_oldest_item(lv_gradient_cache_t * c, void * ctx)
+static lv_grad_t * allocate_item(const lv_grad_dsc_t * g, lv_coord_t w, lv_coord_t h)
 {
     lv_coord_t size = g->dir == LV_GRAD_DIR_HOR ? w : h;
     lv_coord_t map_size = LV_MAX(w, h); /* The map is being used horizontally (width) unless
                                            no dithering is selected where it's used vertically */
-    size_t req_size = sizeof(lv_gradient_cache_t) + map_size * sizeof(lv_color_t)
+
+    size_t req_size = ALIGN(sizeof(lv_grad_t)) + ALIGN(map_size * sizeof(lv_color_t));
 #if _DITHER_GRADIENT
-                      + size * sizeof(lv_color32_t)
+    req_size += ALIGN(size * sizeof(lv_color32_t));
 #if LV_DITHER_ERROR_DIFFUSION == 1
-                      + w * sizeof(lv_scolor24_t)
+    req_size += ALIGN(w * sizeof(lv_scolor24_t));
 #endif
 #endif
-                      ;
+
     size_t act_size = (size_t)(grad_cache_end - LV_GC_ROOT(_lv_grad_cache_mem));
-    if(req_size + act_size > grad_cache_size) {
+    lv_grad_t * item = NULL;
+    if(req_size + act_size < grad_cache_size) {
+        item = (lv_grad_t *)grad_cache_end;
+        item->not_cached = 0;
+    }
+    else {
         /*Need to evict items from cache until we find enough space to allocate this one */
-        if(req_size > grad_cache_size) {
-            LV_LOG_WARN("Gradient cache too small, failed to allocate");
-            return NULL; /*No magic here, if the empty cache is still too small*/
+        if(req_size <= grad_cache_size) {
+            while(act_size + req_size > grad_cache_size) {
+                uint32_t oldest_life = UINT32_MAX;
+                iterate_cache(&find_oldest_item_life, &oldest_life, NULL);
+                iterate_cache(&kill_oldest_item, &oldest_life, NULL);
+                act_size = (size_t)(grad_cache_end - LV_GC_ROOT(_lv_grad_cache_mem));
+            }
+            item = (lv_grad_t *)grad_cache_end;
+            item->not_cached = 0;
         }
-        while(act_size + req_size < grad_cache_size) {
-            uint32_t oldest_life = UINT32_MAX;
-            iterate_cache(&find_oldest_item_life, &oldest_life, NULL);
-            iterate_cache(&kill_oldest_item, &oldest_life, NULL);
-            act_size = (size_t)(grad_cache_end - LV_GC_ROOT(_lv_grad_cache_mem));
+        else {
+            /*The cache is too small. Allocate the item manually and free it later.*/
+            item = lv_mem_alloc(req_size);
+            LV_ASSERT_MALLOC(item);
+            if(item == NULL) return NULL;
+            item->not_cached = 1;
         }
-        /*Ok, now we have space to allocate*/
     }
-    lv_gradient_cache_t * item = (lv_gradient_cache_t *)grad_cache_end;
+
     item->key = compute_key(g, size, w);
     item->life = 1;
     item->filled = 0;
     item->alloc_size = map_size;
     item->size = size;
-    item->map = (lv_color_t *)(grad_cache_end + sizeof(*item));
+    if(item->not_cached) {
+        uint8_t * p = (uint8_t *)item;
+        item->map = (lv_color_t *)(p + (ALIGN(sizeof(*item))));
```

Comment:
I don't think this is required. Seems like the PR has shown it isn't. 
That's because sizeof(item) is aligned already.

Reply:
It's not sure. E.g. -Os can pack the structs automatically, so its size can be e.g. 27 bytes

Reply:
No that's not possible. Else, if you had an array of those structures, the 2nd element in the array would cause unaligned access. The compiler must align the structure field to *at least* the least constrained field. It can pad the structure size (so if you do sizeof(*item) on a 27 bytes structure, it'll be 28 not 27, even if only the first 27 bytes are used in the structure).

Reply:
In all cases, I don't think it's an issue, since if you ALIGN something that's already aligned, it'll give you the same number. It just make the code harder to follow.

Reply:
> No that's not possible. Else, if you had an array of those structures, the 2nd element in the array would cause unaligned access. 

GCC is a little too smart sometimes. I don't know if it happens specifically in this case, but if it knows the target CPU (e.g. Cortex-M7) can support unaligned accesses, and you don't explicitly tell it not to use them with `-mno-unaligned-access`, I have absolutely seen it generate code that uses unaligned accesses like this.

Reply:
I remember that we had issues with GCC packing structs that it wasn't able to read later. That's why we have introduced `LV_ATTRIBUTE_MEM_ALIGN` in `lv_conf.h`.

My experience is that there are too many exotic compilers and hardware so we should do all we can to support special cases too. Especially if it's so simple.

Reply:
If the CPU can support unaligned access, then there's no alignment constraint anymore, right?
 
Damn, this makes my head hurt. You are right, there's the standard on one side with all its rules and there's GCC that bending them to its will and its strong knowledge of each platform. I'm 99% sure that because of "array like" access, structs must be aligned. Even an empty struct has a non zero sizeof because of this, the second element's address must be different than the former.

I don't know all architecture LVGL builds upon. But on ARM and ESP32 (which are both platform with alignment constraint I know about), their rules are quite simple, to access a 16 bits variable, it must be on (at least) a 16 bits boundary, and so on.
So, on those platform, the `ALIGN()` above will do nothing  since it'll happen for both pointers if the item is aligned (which is ensured by my initial PR's ALIGN usage). The last map (`error_acc` might be problematic since its 24 bits wide), but since it's the last, it won't push the other map to unaligned positions. This last map might require alignment to 32 bits or 16 bits, I don't know, probably need testing on a platform where alignment is an issue.

Reply:
@kisvegabor If I guess correctly, you are using a ARM cpu (a mac ?) so you can assert my claims by removing those `ALIGN` lines and it should still work.

Please notice that keeping them, at worst, waste memory since any 16bits variable (`lv_color_t` get pushed to a 32bits or 64bits boundary). So at best, we are wasting 6 bytes per gradient and 12 bytes at worst for this *maybe* useless alignment constraint. Since we can have many of those when animating/occluding/etc... it might worth double checking to avoid this waste.